format/spec.md

```diff
@@ -313,123 +313,81 @@ The set of metadata columns is:
-and explicitly written when the row is moved to a new file.
-* `_last_update` the sequence number of the commit which last updated this row. The value is computed via inheritance for 
-rows in their original file or in files where the row was modified.
+* `_row_id` a unique long for every row. Assigned via inheritance when a row is added and the existing value is explicitly written when the row is written to a new file.
+* `_last_update` the sequence number of the commit that last updated a row. The value is inherited when a row is added or modified and the existing value is explicitly written when the row is written to a different data file but not modified.
 
-Commits with `row-lineage` enabled are not allowed to include any [Equality Deletes](#equality-delete-files).
+These fields are assigned and updated by inheritance because the commit sequence number and starting row ID are not assigned until the snapshot is successfully committed. Inheritance is used to allow writing data and manifest files before values are known so that it is not necessary to rewrite data and manifest files when an optimistic commit is retried.
 
-Implementations writing to tables where `row-lineage` is enabled must populate several additional 
-fields in the metadata and propagate row information from existing and updated. 
+When row lineage is enabled, new snapshots cannot include [Equality Deletes](#equality-delete-files). Row lineage is incompatible with equality deletes because lineage values must be maintained, but equality deletes are used to avoid reading existing data before writing changes.
 
-#### Metadata Propagation
 
-Creating a new commit when `row-lineage` is enabled requires
-* Setting the new snapshot's `first-row-id` field to the previous table metadata's `last-row-id` field
-* Setting `first_row_id` in new data manifest entries in the new manifest list to `first-row-id` field 
-from the snapshot plus the sum of all `added_rows_count` in previously listed new data manifest entries.
-* Setting `first_row_id` in new `data_file` entries  to null and in existing `data_file` structs to their original
-or computed value (if previously null) of `first_row_id`
-* Incrementing the new metadata's `last-row-id` field by the number of new rows added to the table by the commit.
+#### Row lineage assignment
 
-When reading a `data_file` manifest entry with a null `first_row_id`, the value is calculated as the 
-sum of the `total_records` of every previous added `data_file` entry summed with the`first_row_id` of the manifest. 
+Row lineage fields are written when row lineage is enabled. When not enabled, row lineage fields (`_row_id` and `_last_update`) must not be written to data files. The rest of this section applies when row lineage is enabled.
 
+When a row is added or modified, the `_last_update` field is set to `null` so that it is inherited when reading. Similarly, the `_row_id` field for an added row is set to `null` and assigned when reading.
```

Comment:
This requires always adding the fields when writing and I was convinced that this is probably not necessary for what we are doing. If we don't have this requirement it becomes very easy for writers which generate parquet files without these fields to participate in row lineage and only metadata writers need to be modified (except in the case of row level operations). The other benefit is that files can be added to the table from another source (like hive) and still ahve row lineage work.